pseudocode-defns/pseudocode_defns.tex

```diff
@@ -0,0 +1,490 @@
+
+\questionr{Should we have a general definition for ``floats'' (and ``integers''?), or is it sufficiently understood what a float is in general?}
+
+\section{Domains}
+
+$\sfi$ A \emph{data domain} is a representation of the set of values on which a metric or function can operate. For example, if a function accepts inputs from the domain \texttt{IntervalDomain(1:u32,17:u32)}, this means that the function can take any input value \texttt{v} of type \texttt{u32} such that \texttt{1 <= v} and \texttt{v <= 17}.
+
+\questionr{Should anything else be included in this introduction to ``Domains''?}
+
+\begin{definition}[\texttt{AllDomain}]$\sfi$
+\texttt{AllDomain(T)} is the data domain consisting of all values of type $\T$. 
+\end{definition}
+
+For example, \texttt{AllDomain(u32)} is the domain of all values of type \texttt{u32}.
+
+\begin{definition}[\texttt{IntervalDomain}] \sfi
+For a type \texttt{T} that has a total ordering (\texttt{T} has trait \texttt{TotalOrd}) and for values \texttt{L <= U} of type \texttt{T}, \texttt{IntervalDomain(L:T, U:T)} is the domain that contains all values \texttt{v} of type \texttt{T} such that \texttt{L <= v} and \texttt{v <= U}.
+    
+Note that, because both \texttt{L} and \texttt{U} are of type \texttt{T}, there is no need to explicitly pass \texttt{T}; the type \texttt{T} can be inferred. \texttt{IntervalDomain} is defined on any type that implements the trait \texttt{TotalOrd}.\footnote{As of June 28, the OpenDP library requires the weaker condition of partial ordering (implements \texttt{PartialOrd}) instead.}
+\end{definition}
+
+For example, \texttt{IntervalDomain(1:u32, 17:u32)} corresponds to a domain that contains all the \texttt{u32} values \texttt{v} such that \texttt{1 <= v} and \texttt{v <= 17}.
+
+\begin{definition}[\texttt{SizedDomain}] $\sfi$
+    \texttt{SizedDomain(D)} is a domain that specifies the length of the enclosed domain \texttt{D}. The length of elements of the enclosed domain is specified with an element of type \texttt{usize}.
+\end{definition}
+
+For example, \texttt{SizedDomain(VectorDomain(AllDomain(u32), n)} corresponds to the domain of all vectors of length $\texttt{n}$ and elements of type \texttt{u32}.
+
+\begin{definition}[\texttt{VectorDomain}]
+\texttt{VectorDomain(D)} is the domain of all vectors of elements drawn from domain \texttt{D}.
+\end{definition}
+
+For example, \texttt{VectorDomain(AllDomain(u32))::Carrier} is \texttt{Vec(u32)}.
+
+\horizline
+
+\subsection{Notes, todos, questions}
+\todonei{Add clampable domain (\texttt{ClampableDomain}) -- waiting until \texttt{TotalOrd} is fully implemented in the OpenDP library.}
+
+\section{Traits}
+% any traits should be defined here
+
+\begin{definition}[\texttt{Bounded}]
+    A type $\T$ has trait \texttt{Bounded} $\iffText$ $\T$ has some upper bound and some lower bound (some smallest possible value and some largest possible value).
+\end{definition}
+
+\begin{definition}[\texttt{DistanceConstant}]
```

Comment:
I don't follow this definition, or have intuition for what it is capturing

Reply:
I think this is in reference to the "DistanceConstant" definition. Here's a brief explanation of why DistanceConstant is defined how it is.

If a type named TO has the trait "DistanceConstant", it basically means that the type can be used to represent distance between datasets for the stability relation because (1) multiplication and division are defined on the type TO (2) the type TO has a partial ordering so we can test whether some value is less than or equal to another value and (3) the type has the trait InfCast(some_input_type) so that we can ensure that the d_in value represented as a value of type some_input_type can be casted to a value in type TO that is at least as large as d_in (point 3 is important to have in doing the comparison between d_in and d_out)

Reply:
I think we should only use the stability relation for OpenDP metrics that are based on true metrics in the mathematical sense (nonnegative, symmetric, triangle inequality), and these properties should be included in that (though I don't know why we need division - should be fine to use integers like u32).  So we should probably have a trait "TrueMetric" for such metrics, which will require these DistanceConstant properties for the associated type.

Reply:
There is an explanation of DistanceConstant in the code:
https://github.com/opendp/opendp/blob/main/rust/opendp/src/traits.rs#L8-L16

Reply:
Salil, it looks like it is fine to use integers like u32. See [this Rust documentation](https://doc.rust-lang.org/std/ops/trait.Div.html#impl-Div%3Cu32%3E), which shows that it is implemented for u32. (I found definitively that it is implemented in OpenDP by doing "ctrl-b" to find the definition of Div -- I think the link I provided shows that it is implemented, too.)

I think we have kept in mind the idea that division needs to be done "carefully" (e.g. appropriate rounding), so I think the stability relations we've looked at closely should be mathematically sound even when rounding is necessary.